### PR TITLE
More flexible getAvailableFunds

### LINKED ISSUE #546: Fix code scanning alert - Calls inside a loop
Tracking issue for:
- [ ] https://github.com/bosonprotocol/boson-protocol-contracts/security/code-scanning/66

### Description:

`getAvailableFunds` is a view function that returns the information about all funds that entity can withdraw. It accepts entity id as input argument and returns a list of structs containing “tokenAddress”, “tokenName” and “availableAmount”. For efficiency purposes "tokenName" is not stored  in the contract, but is fetched it on the fly every time getAvailableFunds is invoked. This is an external call to token contract. If method “name” does not exist on that contract (possible since it’s not part of ERC20 standard) or it reverts for some reason, it is handled and generic “Token name unspecified” is used in place of "tokenName" and the call still succeeds.

Although reverts are handled, an attacker can still try to break `getAvailableFunds` by making `name` consume all available resources.
Simplest way to achieve this is to make a custom token contract with “name” with an unbounded while loop inside. Although one malicious token does not immediatly break `getAvailableFunds`, combination of multiple does the trick.

`getAvailableFunds` is an external function and its purpose is to help users (or dApps) to easily retrieve all information needed to supply in `withdrawFunds` whenever they want to withdraw (part of) their funds. Malfunction of `getAvailableFunds` therefore presents the biggest issue to dApps.  
On the other hand, protocol is not severly affected. `getAvailableFunds` can be broken for individual users, but this does not affect others that don’t have malicious tokens on their token list.
Even if getAvailableFunds does not work, all funds are still safe and available to withdraw. It’s true that users cannot get the information about the maximum amount they can withdraw, but “withdrawFunds” still works normally.

### Recommendation:
The easiest way, which does not affect other part of the protocol and interfaces is to set manual limit when `name` is called, for example

```solidity
try IERC20Metadata(tokenAddress).name{gas: 100000}() returns (string memory name) 
```

### PR DESCRIPTION
**Pull request by zajck**
Fixes #546 

And also alleviates the consequences of #322 and #679

### CODE REVIEW

**Comment by mischat** [comment]
and more merge conflicts here @zajck

**Comment by mischat** [comment]
fwiw, I would have totally just removed the "name" field, and make that the client's problem, but i like the solution here nonetheless...

File: contracts/interfaces/handlers/IBosonFundsHandler.sol [hunk 1]
```diff
@@ -72,8 +72,40 @@ interface IBosonFundsHandler is IBosonFundsEvents, IBosonFundsLibEvents {
...
+     *
+     * @param _entityId - id of entity for which availability of funds should be checked
+     * @return tokenList - list of token addresses
+     */
+    function getTokenList(uint256 _entityId) external view returns (address[] memory tokenList);
+
+    /**
+     * @notice Returns list of addresses for which the entity has funds available.
+     *
+     * @param _entityId - id of entity for which availability of funds should be checked
```
Comment (anajuliabit) [inline contracts/interfaces/handlers/IBosonFundsHandler.sol:87]: add _limit and _offset parameters here

File: contracts/protocol/facets/FundsHandlerFacet.sol [hunk 2]
```diff
@@ -149,24 +149,75 @@ contract FundsHandlerFacet is IBosonFundsHandler, ProtocolBase {
...
+     * @return tokenList - list of token addresses
+     */
+    function getTokenList(uint256 _entityId) external view override returns (address[] memory tokenList) {
+        return protocolLookups().tokenList[_entityId];
+    }
+
+    /**
+     * @notice Returns list of addresses for which the entity has funds available.
+     *
+     * @param _entityId - id of entity for which availability of funds should be checked
```
Comment (anajuliabit) [inline contracts/protocol/facets/FundsHandlerFacet.sol:166]: add _limit and _offset parameters here

File: contracts/protocol/facets/FundsHandlerFacet.sol [hunk 3]
```diff
@@ -189,6 +240,8 @@ contract FundsHandlerFacet is IBosonFundsHandler, ProtocolBase {
             availableFunds[i].tokenName = tokenName;
             availableFunds[i].availableAmount = entityFunds[tokenAddress];
         }
+
+        return availableFunds;
```
Comment (anajuliabit) [inline contracts/protocol/facets/FundsHandlerFacet.sol:244]: no need for this return exist

**Review by anajuliabit — CHANGES_REQUESTED** [review]
Looks good, just some minor suggestions

**Review by anajuliabit — APPROVED** [review]

**Review by mischat — APPROVED** [review]

**Pull request merged**